[CBRD-26270] Fix ha constraint

youngjun9072 opened the pull request:
[CBRD-26270](http://jira.cubrid.org/browse/CBRD-26270)

### Purpose

누락된 사용자 스펙을 추가하는 이슈이며 다음 제약사항 구현이 추가됩니다.
- HA 모드에서 CREATE TABLE... 시 복제키가 없는 복제 테이블은 생성 금지
- HA 모드에서 ALTER TABLE... 을 이용한 REPLICATION ON <-> OFF 간 상호 변경 금지

추가적으로 리뷰과정에서 누락되었던 코멘트가있어 반영되었습니다. 
- SM_CLASSFLAG_REPLICATION_OFF -> SM_CLASSFLAG_REPLICATION_DATA_OFF로 변경
  데이터 복제를 강조하기위한 이름 변경입니다.
- IS_REPLICATION_ON_OPT -> IS_CREATE_OPT_REPLICATION_ON 로 변경, IS_REPLICATION_ON_NODE -> IS_ALTER_OPT_REPLICATION_ON 로 변경
  "옵션 혹은 노드의 정보를 본다"는 코드적 의미에서 "생성시 혹은 수정시 옵션을 본다"라는 기능적 의미를 반영하기위함입니다.

### Implementation

- CREATE TABLE 시 에러를 발생시키는 조건이 추가됩니다. 조건식은 IS_CREATE_OPT_REPLICATION_ON && !HA_DISABLED && !has_replication_key_constraint (class_obj), 즉, 복제 옵션이 존재하고, HA 모드이며, 복제 키 후보가 존재하지 않을 경우 에러입니다. 
기존에 복제 플레그를 설정하는 코드를 함께 사용하기위해 복제 조건은 else if 구문으로 !HA_DISABLED () && !has_replication_key_constraint (class_obj) 만 비교합니다.
- ALTER TABLE 에서 replication 옵션을 처리할 때, HA 모드일 경우 (!HA_DISABLE)일 경우 에러를 발생시킵니다.

### Remarks

N/A

youngjun9072 committed:
add replication option

youngjun9072 committed:
lint

youngjun9072 committed:
fix typo

youngjun9072 committed:
add "TODO" comment

youngjun9072 committed:
add alter table handling

youngjun9072 committed:
에러 핸들링 추가

youngjun9072 committed:
Refactoring

youngjun9072 committed:
Apply review feedback
- fix indentation
- rename SM_CLASSFLAG_NO_REPLICATION to SM_CLASSFLAG_REPLICATION_OFF and update comment
- check replication status at function entry to avoid unnecessary operations
- adjust error message and parameter from 1 to 0
- add error handling

youngjun9072 committed:
Fix missing user spec and other updates

- Disallow creation of replication tables without replication key in HA mode
- Prevent REPLICATION ON <-> OFF switching via ALTER in HA mode
  (expected complexity increase with foreign key implementation)
- Rename related macros

youngjun9072 committed:
Fix: incorrect replication key detection logic

- Previously, the logic checked constraints and attributes separately.
- This caused a table to be recognized as having a replication key
  even when "UNIQUE" and "NOT NULL" were defined on different columns.
- Updated the logic to require both conditions on the same set of columns.

youngjun9072 committed:
add "alter table ... replication" option

youngjun9072 committed:
lint

youngjun9072 committed:
Code changes

- Changed the location where replication option is handled
- Simplified condition logic to check only default or "on"
- Added explanatory comments

youngjun9072 committed:
리뷰 반영
 - 매크로 이름 변경
 - 불필요 공백 삭제

youngjun9072 committed:
Add macro for alter condition branch

youngjun9072 committed:
add replication opt handling and select catalog

youngjun9072 committed:
replication 저장 및 조회 구현

youngjun9072 committed:
add comment

youngjun9072 committed:
Code modification

- Rename flag from SM_CLASSFLAG_REPLICATION to SM_CLASSFLAG_NO_REPLICATION
- Update return values to reflect the new flag naming
- Add error handling when attempting to alter replication tables in HA mode

youngjun9072 committed:
Macro refactoring and lint

youngjun9072 committed:
lint

youngjun9072 committed:
replication pk only

youngjun9072 committed:
Implementation of unique key insert, delete, and update

youngjun9072 committed:
fix(pgbuf): prevent assertion in pgbuf_is_valid_page during UPDATE/DELETE

youngjun9072 committed:
refectoring

youngjun9072 committed:
lint

H2SU commented on src/query/execute_schema.c:
```diff
@@ -8853,6 +8854,24 @@ execute_create_select_query (PARSER_CONTEXT * parser, const char *const class_na
+      return false;
+    }
+
+  return classobj_has_class_rk_constraint (c);
```
classobj_has_class_rk_constraint 함수에 전달될 c가 NULL이라면 해당 함수에서 false를 반환하는 것으로 보입니다.
간결하게 아래처럼 해도 되지않을까요..?

```suggestion
  return classobj_has_class_rk_constraint (db_get_constraints (class_obj));
```

youngjun9072 replied:
@H2SU 나이스캐치입니다. 말씀하신데로 리팩토링해보니 has_replication_key_constraint()함수도 필요없을것같아 해당함수를 삭제했고, execute에서 바로 classobj_has_class_rk_constraint()를 호출하도록했습니다.

hornetmj commented on src/object/class_object.h:
```diff
@@ -309,7 +309,7 @@ typedef enum
   SM_CLASSFLAG_REUSE_OID = 8,	/* the class can reuse OIDs */
   SM_CLASSFLAG_SUPPLEMENTAL_LOG = 16,	/* reserved flag for supplemental log. */
-  SM_CLASSFLAG_REPLICATION_OFF = 32	/* disables replication of class data (schema still replicated) */
+  SM_CLASSFLAG_REPLICATION_DATA_OFF = 32	/* disables replication of class data (schema still replicated) */
```
```suggestion
  SM_CLASSFLAG_DATA_REPLICATION_OFF = 32	/* disables replication of class data (schema still replicated) */
```

hornetmj replied:
주석들 정렬을 맞춰주시면 좋을 것 같습니다.

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj replied:
<img width="963" height="147" alt="화면 캡처 2025-10-21 165153" src="https://github.com/user-attachments/assets/adf8a90d-511a-495d-8ad3-f8c418427f72" />

위와 같이 확인되고, 새로운 commit이 올라오지 않은 것 같습니다. 확인 부탁해요.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -89,16 +89,17 @@
+ * Used during CREATE TABLE parsing to check table option node (tbl_opt_replication).
  */
-#define IS_REPLICATION_ON_OPT(_opt) \
+#define IS_CREATE_OPT_REPLICATION_ON(_opt) \
```
```suggestion
#define IS_CREATE_STMT_SET_REPL_OPTION(_opt) \
```

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -89,16 +89,17 @@
+ * Used during ALTER TABLE parsing to check replication option node (replication_node).
  */
-#define IS_REPLICATION_ON_NODE(_node) \
+#define IS_ALTER_OPT_REPLICATION_ON(_node) \
```
```suggestion
#define IS_ALTER_STMT_SET_REPL_OPTION(_node) \
```

youngjun9072 replied:
@hornetmj 변경하였습니다.

hornetmj commented on src/object/class_object.c:
```diff
@@ -2183,6 +2183,28 @@ classobj_has_class_unique_constraint (SM_CLASS_CONSTRAINT * constraints)
+ *   constraints(in): constraint list
+ */
+bool
+classobj_has_class_rk_constraint (SM_CLASS_CONSTRAINT * constraints)
```
추후 관련 코드를 키워드로 찾기 쉽도록, 전체적인 구현 용어를 통일하면 좋을 것 같습니다. rk_, replication_key_, repl_key_, rkey_ 등이 후보군이 될 수 있을 것 같습니다.
```suggestion
classobj_has_class_repl_key_constraint (SM_CLASS_CONSTRAINT * constraints)
```

youngjun9072 replied:
@hornetmj  일단 classobj_has_class_repl_key_constraint() 로 반영하였습니다. 키워드에대해서는 대략적으로 기능과 함수들의 구현 반영하도록하겠습니다

hornetmj commented on src/object/class_object.c:
```diff
@@ -2183,6 +2183,28 @@ classobj_has_class_unique_constraint (SM_CLASS_CONSTRAINT * constraints)
+
+  for (c = constraints; c != NULL; c = c->next)
+    {
+      if (SM_IS_CONSTRAINT_UNIQUE_FAMILY (c->type) && sm_has_non_null_attribute (c->attributes))
```
PK의 경우 not null이 보장되기 때문에 sm_has_non_null_attribute() 호출을 안할 수 있도록 다음 조건을 추가하면 좋을 것 같습니다.
```suggestion
      if ((c->type == SM_CONSTRAINT_PRIMARY_KEY) || (SM_IS_CONSTRAINT_UNIQUE_FAMILY (c->type) && sm_has_non_null_attribute (c->attributes)))
```

youngjun9072 replied:
@hornetmj 반영하였습니다. 수정 전, 후 둘다 더블체크인것같긴한데 추후 새로운 매크로를 추가할만한지 검토해보려합니다.

youngjun9072 replied:
@hornetmj 추가로 조건식이 많아 IS_HA_REPLICATION_KEY_CONSTRAINT라는 매크로를 추가했습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -9234,19 +9253,26 @@ do_create_entity (PARSER_CONTEXT * parser, PT_NODE * node)
+
 	  do_flush_class_mop = true;
 	}
+      else if (!HA_DISABLED () && !has_replication_key_constraint (class_obj))
```
if와 else if의 모든 조건들에 !(NOT)이 붙고, 테이블 옵션 설정 검사와 HA 모드에서 복제키 제약이 같은 레벨에서 검사되는 것 보다는 다음 방식이 더 가독성이 좋을 것 같아 제안드립니다.
```
if (IS_CREATE_OPT_REPLICATION_ON (tbl_opt_replication)) // 테이블 옵션 처리 if ~ else
{
    if (!HA_DISABLED () && !has_replication_key_constraint (class_obj)) // HA 모드에서 제약 조건 검사
    {
        ...
    }
}
else
{
    sm_set_class_flag (class_obj, SM_CLASSFLAG_REPLICATION_DATA_OFF, TRUE);
}
```

youngjun9072 replied:
@hornetmj 반영하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -8853,6 +8854,24 @@ execute_create_select_query (PARSER_CONTEXT * parser, const char *const class_na
+ *   class_obj(in): class object pointer
+ */
+bool
+has_replication_key_constraint (DB_OBJECT * class_obj)
```
schema_manager.c에서 sm_class_has_replication_key_constraint()로 정의하는게 어떨까요?

youngjun9072 replied:
@hornetmj 희수님 코멘트에따라 리팩토링해보니 has_replication_key_constraint() 함수를 호출하지 않고 execute에서 바로 classobj_has_class_repl_key_constraint() 함수를 호출이 가능해 해당 함수를 제거하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10684,12 +10710,13 @@ do_alter_change_replication (PARSER_CONTEXT * const parser, PT_NODE * const alte
-      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0);
+      error = ER_REPLICATION_OPTION_CHANGE_IN_HA_MODE;
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0,
+	      "Changing the replication option of a class is not allowed in HA mode.");
```
```suggestion
      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0);
```

youngjun9072 replied:
@hornetmj 반영하였습니다.

hornetmj commented on src/query/execute_schema.c:
```diff
@@ -10684,12 +10710,13 @@ do_alter_change_replication (PARSER_CONTEXT * const parser, PT_NODE * const alte
+      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, error, 0,
+	      "Changing the replication option of a class is not allowed in HA mode.");
 
       return error;
```
함수의 구현 일관성을 위해 goto exit; 후 error를 리턴하도록 하는게 어떨까요?

hornetmj commented on src/transaction/locator_sr.c:
```diff
@@ -8813,7 +8813,6 @@ locator_update_index (THREAD_ENTRY * thread_p, RECDES * new_recdes, RECDES * old
 	  error_code =
 	    repl_log_insert (thread_p, class_oid, oid, LOG_REPLICATION_DATA, RVREPL_DATA_UPDATE, repl_old_key,
 			     (REPL_INFO_TYPE) repl_info->repl_info_type);
-
```
이 공백은 유지하는게 가독성이 좋을 것 같습니다.

youngjun9072 replied:
@hornetmj 반영하였습니다.

youngjun9072 committed:
Apply review feedback
- Renamed function classobj_has_class_rk_constraint() → classobj_has_class_repl_key_constraint()
- Changed the call structure from
  do_create_entity() -> has_replication_key_constraint() -> classobj_has_class_repl_key_constraint()
  to
  do_create_entity() -> classobj_has_class_repl_key_constraint() directly.
  Since classobj_has_class_repl_key_constraint() already returns false when NULL,
  the NULL check in has_replication_key_constraint() is unnecessary.

- Updated several constant and macro names
- Modified the condition in classobj_has_class_repl_key_constraint() to (PK || (UNIQUE_FAMILY && not null unique))
- Reordered the RK constraint and PK option handling logic in do_create_entity()
- Removed redundant error messages and extra spaces
